### PR TITLE
2021 Facelift

### PR DESCRIPTION
**Pull request by tatumalenko**
This project had really good bones, but there were a few issues with how outdated it was so I'm gonna take a crack at it.

Namely, the following enhancements were brought:
- A lot of the dependencies needed an facelift, and some weren't working anymore for some reason (`funktionale`) -- it was removed and Kotlin's built in null-type awareness was used to replace the `Some`/`None` usages and many dependencies were updated.
- A really important feature I needed was to mark the search result messages as read -- that was added as `markAsRead` in the `SearchBuilder`.
- The `SearchBuilderTest` was using both `WordSpec` (`kotest`, formerly known as `kotlintest`) and `@Test` annotations (`junit`), it was hard to follow and with the big update to `kotlintest`, it was hard to read -- `WordSpec` was replaced with `AnnotationSpec` to keep it more consistent with the other numerous tests in the same class.
- It seems like the ssl property set to enabled is needed for my Gmail use case to work properly (connection is never established otherwise) whereas it needed to be disabled for `GreenMail` -- that was added as an option to the `ConnectionInformation` data class along with `debug` option.
- A bit of sprucing up to the `Flags` class was also added and other parts here and there.
- The readme had a few typos and a few formatting inconsistencies -- those were corrected along with edits to the example usage including `markAsRead`, `sslEnabled`, and `debug`.

My plan next is to add a listener capability similar to how `IdleManager` from `javax.mail` works (although not using it specifically since it has a lot of issues).

### CODE REVIEW

**Comment by tatumalenko** [comment]
@SkittishSloth Might need your help with the Travis builds failing, `openjdk11` seems to fix one issue but now something is failing at the first `before_install` step, related to gpg and signing I believe. What I found different from a previous build that passed was that not all the secrets seem to be exported now. Namely, we observe the following differences.

[Before](https://travis-ci.org/github/SlothLabs/kotlin-mail/builds/124333002):
```sh
Setting environment variables from repository settings
$ export GPG_EXECUTABLE=gpg
$ export GPG_PASSPHRASE=[secure]
$ export SONATYPE_USERNAME=[secure]
$ export SONATYPE_PASSWORD=[secure]
$ export GPG_SECRET_KEYS=[secure]
$ export GPG_OWNERTRUST=[secure]
```

[Now](https://travis-ci.org/github/SlothLabs/kotlin-mail/builds/760777721):
```sh
Setting environment variables from repository settings
$ export GPG_EXECUTABLE=gpg
```

Did you remove the secrets by chance after publishing `0.1.0`?

**Comment by SkittishSloth** [comment]
Hey @tatumalenko - I'm sorry for the delay, I didnt see the emails until just now. Thank you very much for giving this old code base some attention; I obviously haven't. I'll try and go over your commits here in a little bit but I wanted to at least chime in here real quick since you've already been waiting a few days.

Regarding the Travis CI secrets, I honestly have no idea how I handled those, and I have no idea how to begin to fix that. I don't even have the machine I used when I was working on this (it was a company laptop, and I haven't worked there for about 4 years now), so I can't even poke around and see if it's there somewhere. That said, I'll try and dig into it and see what I can figure out on my end.

**Comment by tatumalenko** [comment]
Thanks for the heads ups! I tried googling it a bit since I'm not super familiar with Maven nor publishing to maven central, it seems like it's related to artifact signing. The first thing I'd check if I were you is whether the secrets are still stored for all those environment variables I showed you in the screenshot (they should be in your travisCI project settings for secrets/environment variables on their website). Perhaps the secrets aren't valid anymore? Worse case, we could take the opportunity to revamp the workflow to use Github Actions instead of TravisCI with possibly new sonar type and maven central secrets.

**Comment by SkittishSloth** [comment]
I looked earlier; there's values there, but I have no clue if they're valid or not. Migrating to GH Actions is a good idea - that definitely wasn't an option when I set this up way back when.

**Comment by tatumalenko** [comment]
Wow, the [master build](https://travis-ci.org/github/SlothLabs/kotlin-mail/builds/761337445#L2728) had much better luck than the PR one. It still failed but it managed to run the unit tests at least. 

```
Skipping a deployment with the script provider because this is not on the required runtime
Skipping a deployment with the script provider because this is not on the required runtime
Skipping a deployment with the script provider because this is not a tagged commit
$ rvm $(travis_internal_ruby) --fuzzy do ruby -S gem install dpl
Successfully installed dpl-1.10.16
1 gem installed
invalid option "--token="
failed to deploy
```

I can't tell if it's the "skipping deployment..." part or the "invalid option..." that caused the build to fail. What I could do is open a PR to change the values of the jdk version of the deploy task which I forgot to change when I changed the value for the pipeline on line 3 to see if that fixes the "not on required runtime" issue.
